### ISSUE TITLE
Unexpected behavior of command denote-dired-rename-file

### ISSUE THREAD

**Issue by hpgisler** [protesilaos/denote#63]
### Situation
I have a file with name ```20220724T171924--test__psychology_work.org``` and content:
```
#+title: Align the psychology  works
#+date: [2022-06-30 Do]
#+filetags: motivation


```
1. If I open ```dired```, set point to said file and call ```denote-dired-rename-file``` and 
2. enter a new name, like ```Align the motivations of```
3. and select eg. 2 keywords, say ```work,psychology```
4. I get asked whether this is ok (which I answer with ```y```
5. I get to correct diff (red/green) with new title and new keywords, (which I answer with ```y```)
6. After that, I get 3 files after updating my dired buffer (with g)
- #20220724T171924--align-the-motivations-of__psychology_work.org#
- .#20220724T171924--align-the-motivations-of__psychology_work.org -> hp@hu.224559:1658053153
- 20220724T171924--align-the-motivations-of__psychology_work.org
7. when I then do ```org-save-all-org-buffers and again buffer update (g), I've one file left: 
- 20220724T171924--align-the-motivations-of__psychology_work.org
8. opening this files yields:
```
#+title: Align the psychology  works of
#+date: [2022-06-30 Do]
#+filetags: motivation

```

i.e. the file remains unchanged.

### Expected behavior
```
#+title: Align the motivations of
#+date: [2022-06-30 Do]
#+filetags: work psychology

```

### Note
I seem to be unable to reproduce this strange behavior with a simpler file...

### Versions
(version) yields: "GNU Emacs 28.1 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.33, cairo version 1.17.6)
Denote Version (GitHub): 61e672f main origin/main Replace needless 'when-let*' with 'when-let'

**Comment by protesilaos** [comment]
Hello @hpgisler!

This is code I changed earlier today: https://github.com/protesilaos/denote/issues/60

There may have been a regression.  Will check and report back to you.

**Comment by protesilaos** [comment]
By the way, the extra files look like the backups Emacs generates automatically.  This is because the buffer is not saved.

**Comment by protesilaos** [comment]
I am not sure what is happening here.  It is not related to the change I did before.

If I have this front matter:

```
#+title: Align the psychology  works of
#+date: [2022-06-30 Do]
#+filetags: motivation

```

And do `M-:` to evaluate `(denote-retrieve--search (buffer-file-name) denote-retrieve--title-front-matter-key-regexp)` its return value is `"Align the psychology  works"`.

But if I make a tiny edit in the title and either keep it or undo it, the above returns the correct value.

This needs more debugging.

**Comment by hpgisler** [comment]
Hello Prot,

I believe I see the problem: In function:

```lisp
(defun denote-dired--rewrite-front-matter (file title keywords)
  "Rewrite front matter of note after `denote-dired-rename-file'.
The FILE, TITLE, and KEYWORDS are passed from the renaming
command and are used to construct a new front matter block if
appropriate."
  (when-let* ((denote-dired--edit-front-matter-p file)
              (id (denote-retrieve--filename-identifier file))
              (date (denote-retrieve--value-date file)))
    (let ((old-title (denote-retrieve--value-title file))
          (old-keywords (denote-retrieve--value-keywords file))
          (new-title title)
          (new-keywords (denote--file-meta-keywords
                         keywords (denote-dired--filetype-heuristics file))))
      (with-current-buffer (find-file-noselect file)
        (when (y-or-n-p (format
                         "Replace front matter?\n-%s\n+%s\n\n-%s\n+%s"
                         (propertize old-title 'face 'error)
                         (propertize new-title 'face 'success)
                         (propertize old-keywords 'face 'error)
                         (propertize new-keywords 'face 'success)))
          (save-excursion
            (goto-char (point-min))
            (search-forward old-title nil t 1)
            (replace-match (concat "\\1" new-title) t)
            (goto-char (point-min))
            (search-forward old-keywords nil t 1)
            (replace-match (concat "\\1" new-keywords) t)))))))
```

In the last part (```save-excursion```), there are two independent search&replace ops:
1. old-title --> new-title
2. old-keywords --> new-keywords

Now, if - as is the case in my example - the new-title contains a word which happens to be equal to the old-keyword, the following happens:
1. old-title --> new-title (with old-keyword somewhere in the title)
2. the second search&replace will now replace the title's old-keyword with the new-keyword (only), because the search&replace will regard only the first match
3. the old-keyword thus will remain unchanged and the new-title will get mixed up

**Comment by protesilaos** [comment]
> I believe I see the problem: In function:

Yes, what you explain here is correct.  We need to include a search for the key as well.  For example (not tested yet):

```diff
 denote-dired.el | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/denote-dired.el b/denote-dired.el
index d990169..ca9d033 100644
--- a/denote-dired.el
+++ b/denote-dired.el
@@ -302,11 +302,15 @@ (defun denote-dired--rewrite-front-matter (file title keywords)
                          (propertize new-title 'face 'success)
                          (propertize old-keywords 'face 'error)
                          (propertize new-keywords 'face 'success)))
+          ;; TODO 2022-07-25: Perhaps we can make
+          ;; `denote-retrieve--value-title' more efficient?
           (save-excursion
             (goto-char (point-min))
+            (denote-retrieve--value-title file t)
             (search-forward old-title nil t 1)
             (replace-match (concat "\\1" new-title) t)
             (goto-char (point-min))
+            (denote-retrieve--value-keywords file t)
             (search-forward old-keywords nil t 1)
             (replace-match (concat "\\1" new-keywords) t)))))))
 
```

Though I hadn't gone there yet because I was encountering a strange behaviour in the internal function that performs the search, as explained here: https://github.com/protesilaos/denote/issues/63#issuecomment-1193364458.

**Comment by protesilaos** [comment]
Just pushed commit `36440d7`:

```
commit 36440d7b7dfb975e8a3c87e52208d8a59c213ed3
Author: Protesilaos Stavrou <info@protesilaos.com>
Date:   Mon Jul 25 03:50:17 2022 +0300

    Refine front matter rewrite
    
    This should avoid false positives by anchoring the search at the
    relevant front matter key.
    
    The code is broken into smaller parts for easier hacking.
    
    This change should address the bug with false positives reported by
    hpgisler in issue 63 on the GitHub mirror:
    <https://github.com/protesilaos/denote/issues/63>.
```

```diff
  denote-dired.el    | 28 +++++++++++++++++++++-------
 denote-retrieve.el |  3 +++
 2 files changed, 24 insertions(+), 7 deletions(-)

diff --git a/denote-dired.el b/denote-dired.el
index d990169..8ece6de 100644
--- a/denote-dired.el
+++ b/denote-dired.el
@@ -282,6 +282,25 @@ (defun denote-dired--edit-front-matter-p (file)
          ;; Heuristic to check if this is one of our notes
          (string= (expand-file-name default-directory) (denote-directory)))))
 
+;; FIXME 2022-07-25: We should make the underlying regular expressions
+;; that `denote-retrieve--value-title' targets more refined, so that we
+;; capture eveyrhing at once.
+(defun denote-dired--rewrite-title (file old new)
+  "Rewrite OLD front matter of title to NEW value in FILE."
+  (save-excursion
+    (goto-char (point-min))
+    (re-search-forward denote-retrieve--title-front-matter-key-regexp nil t 1)
+    (search-forward old nil t 1)
+    (replace-match (concat "\\1" new) t)))
+
+(defun denote-dired--rewrite-keywords (file old new)
+  "Rewrite OLD front matter of keywords to NEW value in FILE."
+  (save-excursion
+    (goto-char (point-min))
+    (re-search-forward denote-retrieve--keywords-front-matter-key-regexp nil t 1)
+    (search-forward old nil t 1)
+    (replace-match (concat "\\1" new) t)))
+
 (defun denote-dired--rewrite-front-matter (file title keywords)
   "Rewrite front matter of note after `denote-dired-rename-file'.
 The FILE, TITLE, and KEYWORDS are passed from the renaming
@@ -302,13 +321,8 @@ (defun denote-dired--rewrite-front-matter (file title keywords)
                          (propertize new-title 'face 'success)
                          (propertize old-keywords 'face 'error)
                          (propertize new-keywords 'face 'success)))
-          (save-excursion
-            (goto-char (point-min))
-            (search-forward old-title nil t 1)
-            (replace-match (concat "\\1" new-title) t)
-            (goto-char (point-min))
-            (search-forward old-keywords nil t 1)
-            (replace-match (concat "\\1" new-keywords) t)))))))
+          (denote-dired--rewrite-title file old-title new-title)
+          (denote-dired--rewrite-keywords file old-keywords new-keywords))))))
 
 (defun denote-dired--add-front-matter (file title keywords id)
   "Add front matter to the beginning of FILE.
diff --git a/denote-retrieve.el b/denote-retrieve.el
index fd21df0..0e6ef21 100644
--- a/denote-retrieve.el
+++ b/denote-retrieve.el
@@ -33,6 +33,9 @@ ;;; Code:
 (require 'denote)
 (require 'xref)
 
+;; FIXME 2022-07-25: We must review these.  When we need to move in the
+;; buffer, we should capture both the key and its value at once.  Check
+;; `denote-dired--rewrite-front-matter'.
 (defconst denote-retrieve--title-front-matter-key-regexp
   "^\\(?:#\\+\\)?\\(?:title\\)\\s-*[:=]"
   "Regular expression for title key.")
```

**Comment by protesilaos** [comment]
I did make the final change in commit `47073aa`.  Please check.

**Comment by hpgisler** [comment]
Hello Prot,
Thank you very much! this works now for me as expeced.
Take care (and make sure you get enough sleep!)
Best.
HP.

**Comment by protesilaos** [comment]
> Thank you very much! this works now for me as expeced.

You are welcome!

> Take care (and make sure you get enough sleep!)

You are right!  Normally I do, though these last few days have been the exception.

I just released version `0.4.0`: <https://protesilaos.com/codelog/2022-07-25-denote-0-4-0/>.  I think [your other issue](https://github.com/protesilaos/denote/issues/62) with `when-let*` will be solved as well, but let's wait to see.

I am closing this now.

Thanks again for your time and feedback!
